File: samples/GNWrapperAdSample/kotlin_app/src/main/java/jp/co/geniee/gnwrapperadsample/MainActivity.java
package jp.co.geniee.gnwrapperadsample;

import android.os.Build;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.webkit.WebView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.doubleclick.AppEventListener;
import com.google.android.gms.ads.doubleclick.PublisherAdRequest;
import com.google.android.gms.ads.doubleclick.PublisherAdView;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings;

import org.prebid.mobile.PrebidMobile;
import org.prebid.mobile.addendum.AdViewUtils;
import org.prebid.mobile.addendum.PbFindSizeError;

import jp.co.geniee.gnwrapperadsdk.GNWrapperAdSDK;
import jp.co.geniee.gnwrapperadsdk.banner.GNWrapperAdBanner;
import jp.co.geniee.gnwrapperadsdk.enums.GNErrorCode;
import jp.co.geniee.gnwrapperadsdk.enums.GNLogLevel;
import jp.co.geniee.gnwrapperadsdk.listeners.GNWrapperAdBannerListener;
import jp.co.geniee.gnwrapperadsdk.params.GNCustomTargetingParams;

public class MainActivity extends AppCompatActivity implements GNWrapperAdBannerListener, AppEventListener {
    private static final String TAG = "MainActivity";
    private static final String GN_WRAPPER_CONFIG = "GNWrapperConfig_Android";
    private PublisherAdView mPublisherAdView;
    private FirebaseRemoteConfig mFirebaseRemoteConfig;
    private GNWrapperAdBanner gnWrapperAdBanner;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        GNWrapperAdSDK.setLogLevel(GNLogLevel.DEBUG);
        GNWrapperAdSDK.setTestMode(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(true);
        }

        final RelativeLayout constraintLayout = findViewById(R.id.constraintLayout);

        PrebidMobile.setApplicationContext(getApplicationContext());

        mFirebaseRemoteConfig = FirebaseRemoteConfig.getInstance();
        FirebaseRemoteConfigSettings configSettings = new FirebaseRemoteConfigSettings.Builder()
                .setMinimumFetchIntervalInSeconds(3600)
                .build();
        mFirebaseRemoteConfig.setConfigSettingsAsync(configSettings);
        mFirebaseRemoteConfig.setDefaultsAsync(R.xml.remote_config_defaults);
        mFirebaseRemoteConfig.fetchAndActivate()
                .addOnCompleteListener(this, new OnCompleteListener<Boolean>() {
                    @Override
                    public void onComplete(@NonNull Task<Boolean> task) {
                        if (task.isSuccessful()) {
                            Log.d(TAG, "FirebaseRemoteConfig is Success");
                            String gnGNRemoteConfigValue = mFirebaseRemoteConfig.getString(GN_WRAPPER_CONFIG);
                            Log.d(TAG, "gnGNRemoteConfigValue: " + gnGNRemoteConfigValue);
                            gnWrapperAdBanner.initAndLoad(gnGNRemoteConfigValue);
                        } else {
                            Log.d(TAG, "FirebaseRemoteConfig is failed Exception: " + task.getException());
                        }
                    }
                });

        mPublisherAdView = new PublisherAdView(MainActivity.this);
        mPublisherAdView.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                Log.d(TAG, "onAdLoaded");
                mPublisherAdView.setVisibility(View.VISIBLE);
                AdViewUtils.findPrebidCreativeSize(mPublisherAdView, new AdViewUtils.PbFindSizeListener() {
                    @Override
                    public void success(int width, int height) {
                        Log.d(TAG, "Resize Ad: width: " + width + " height: " + height);
                        mPublisherAdView.setAdSizes(new AdSize(width, height));
                    }
                    @Override
                    public void failure(@NonNull PbFindSizeError error) {
                        Log.d("MyTag", "error: " + error);
                    }
                });
                // Code to be executed when an ad finishes loading.
            }

            @Override
            public void onAdFailedToLoad(LoadAdError adError) {
                Log.d(TAG, "onAdFailedToLoad");
                // Code to be executed when an ad request fails.
            }

            @Override
            public void onAdOpened() {
                Log.d(TAG, "onAdOpened");
                // Code to be executed when an ad opens an overlay that
                // covers the screen.
            }

            @Override
            public void onAdClicked() {
                Log.d(TAG, "onAdClicked");
                // Code to be executed when the user clicks on an ad.
            }

            @Override
            public void onAdClosed() {
                Log.d(TAG, "onAdClosed");
                // Code to be executed when the user is about to return
                // to the app after tapping on an ad.
            }
        });

        mPublisherAdView.setAppEventListener(this);
        mPublisherAdView.setId(View.generateViewId());
        constraintLayout.addView(mPublisherAdView);

        gnWrapperAdBanner = new GNWrapperAdBanner(this, this);
        gnWrapperAdBanner.setId(View.generateViewId());

        gnWrapperAdBanner.setLayoutParams(new LinearLayout.LayoutParams(convertDp2Px(320), convertDp2Px(50)));

        constraintLayout.addView(gnWrapperAdBanner);


        //setCenterConstraintSet(constraintLayout);
    }


    private void setCenterConstraintSet(ConstraintLayout constraintLayout) {

        ConstraintSet constraintSet = new ConstraintSet();
        constraintSet.clone(constraintLayout);

        constraintSet.constrainWidth(mPublisherAdView.getId(),
                convertDp2Px(300));

        constraintSet.constrainHeight(mPublisherAdView.getId(),
                convertDp2Px(250));

        constraintSet.constrainWidth(gnWrapperAdBanner.getId(),
                convertDp2Px(300));

        constraintSet.constrainHeight(gnWrapperAdBanner.getId(),
                convertDp2Px(250));

        constraintSet.connect(
                mPublisherAdView.getId(),
                ConstraintSet.BOTTOM,
                ConstraintSet.PARENT_ID,
                ConstraintSet.BOTTOM,
                0);

        constraintSet.connect(
                mPublisherAdView.getId(),
                ConstraintSet.LEFT,
                ConstraintSet.PARENT_ID,
                ConstraintSet.LEFT,
                0);

        constraintSet.connect(
                mPublisherAdView.getId(),
                ConstraintSet.RIGHT,
                ConstraintSet.PARENT_ID,
                ConstraintSet.RIGHT,
                0);

        constraintSet.connect(
                mPublisherAdView.getId(),
                ConstraintSet.TOP,
                ConstraintSet.PARENT_ID,
                ConstraintSet.TOP,
                0);

        constraintSet.connect(
                gnWrapperAdBanner.getId(),
                ConstraintSet.BOTTOM,
                ConstraintSet.PARENT_ID,
                ConstraintSet.BOTTOM,
                0);

        constraintSet.connect(
                gnWrapperAdBanner.getId(),
                ConstraintSet.LEFT,
                ConstraintSet.PARENT_ID,
                ConstraintSet.LEFT,
                0);

        constraintSet.connect(
                gnWrapperAdBanner.getId(),
                ConstraintSet.RIGHT,
                ConstraintSet.PARENT_ID,
                ConstraintSet.RIGHT,
                0);

        constraintSet.connect(
                gnWrapperAdBanner.getId(),
                ConstraintSet.TOP,
                ConstraintSet.PARENT_ID,
                ConstraintSet.TOP,
                0);

        constraintSet.applyTo(constraintLayout);
        setContentView(constraintLayout);
    }

    public int convertDp2Px(float dp){
        DisplayMetrics metrics = this.getResources().getDisplayMetrics();
        return (int) (dp * metrics.density);
    }

    @Override
    public void onComplete(String adUnitId, GNCustomTargetingParams gnCustomTargetingParams) {
        if (mPublisherAdView.getAdUnitId() == null) {
            mPublisherAdView.setAdUnitId(adUnitId);
            mPublisherAdView.setAdSizes(AdSize.BANNER);
        }

        PublisherAdRequest.Builder adRequestBuilder = new PublisherAdRequest.Builder();
        for (String key: gnCustomTargetingParams.keySet()) {
            Log.d(TAG, "key: " + key + " value: " + gnCustomTargetingParams.get(key));
            adRequestBuilder.addCustomTargeting(key, gnCustomTargetingParams.get(key));
        }

        PublisherAdRequest adRequest = adRequestBuilder.build();

        mPublisherAdView.loadAd(adRequest);
    }

    @Override
    public void onError(GNErrorCode gnErrorCode) {
        Log.w(TAG, "GNWrapperError: " + gnErrorCode.getMessage());
    }

    @Override
    public void onAppEvent(String s, String s1) {
        if(s.equals("pubmaticdm")) {
            Log.d(TAG, " GNWrapperAd.isShowInSDKView: " +  gnWrapperAdBanner.isShowInSDKView());
            if (gnWrapperAdBanner.isShowInSDKView()) {
                mPublisherAdView.setVisibility(View.INVISIBLE);
                gnWrapperAdBanner.show();
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mPublisherAdView != null) {
            mPublisherAdView.destroy();
        }
        if (gnWrapperAdBanner != null) {
            gnWrapperAdBanner.destroy();
        }
    }
}

File: samples/GNWrapperAdSample/kotlin_app/src/main/java/jp/co/geniee/gnwrapperadsample/MainActivity.kt
package jp.co.geniee.gnwrapperadsample

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.ads.*
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import com.google.android.gms.ads.admanager.AppEventListener
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings
import jp.co.geniee.gnwrapperadsdk.GNWrapperAdSDK
import jp.co.geniee.gnwrapperadsdk.banner.GNWrapperAdBanner
import jp.co.geniee.gnwrapperadsdk.enums.GNErrorCode
import jp.co.geniee.gnwrapperadsdk.enums.GNLogLevel
import jp.co.geniee.gnwrapperadsdk.listeners.GNWrapperAdBannerListener
import jp.co.geniee.gnwrapperadsdk.params.GNCustomTargetingParams
import org.prebid.mobile.PrebidMobile
import org.prebid.mobile.addendum.AdViewUtils
import org.prebid.mobile.addendum.AdViewUtils.PbFindSizeListener
import org.prebid.mobile.addendum.PbFindSizeError


class MainActivity : AppCompatActivity(), GNWrapperAdBannerListener, AppEventListener {
    private var adManagerAdView: AdManagerAdView? = null
    private var firebaseRemoteConfig: FirebaseRemoteConfig? = null
    private var gnWrapperAdBanner: GNWrapperAdBanner? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        MobileAds.initialize(this) { }
        /*
        //  TODO: If you want to debug your code, please refer to the URL below and enter your Test device ID.
        //  EN: https://developers.google.com/ad-manager/mobile-ads-sdk/android/test-ads#enable_test_devices
        //  JP: https://developers.google.com/ad-manager/mobile-ads-sdk/android/test-ads?hl=ja#enable_test_devices
        */
        val testDeviceIds = listOf("ADD_YOUR_TEST_DEVICE_ID")
        val configuration = RequestConfiguration.Builder().setTestDeviceIds(testDeviceIds).build()
        MobileAds.setRequestConfiguration(configuration)

        // Change the GNLogLevel if you want logcat to display the GNWrapperAdSDK debug code
        GNWrapperAdSDK.setLogLevel(GNLogLevel.INFO)

        /*
        // Make this setting to run the GNWrapperAdSDK in test mode.
        // TODO: Be sure to remove it when releasing
         */
        GNWrapperAdSDK.setTestMode(true)

        PrebidMobile.setApplicationContext(applicationContext)

        val adView = findViewById<FrameLayout>(R.id.ad_view_layout)

        adManagerAdView = AdManagerAdView(this)
        adManagerAdView!!.adListener = object : AdListener() {
            override fun onAdLoaded() {
                Log.d(TAG, "onAdLoaded")

                adManagerAdView!!.visibility = View.VISIBLE
                AdViewUtils.findPrebidCreativeSize(adManagerAdView, object : PbFindSizeListener {
                    override fun success(width: Int, height: Int) {
                        Log.d(TAG, "Resize Ad: width: $width height: $height")
                        adManagerAdView!!.setAdSizes(AdSize(width, height))
                    }

                    override fun failure(error: PbFindSizeError) {
                        Log.d("MyTag", "error: $error")
                    }
                })
            }

            override fun onAdFailedToLoad(adError: LoadAdError) {
                Log.d(TAG, "onAdFailedToLoad")
            }

            override fun onAdOpened() {
                Log.d(TAG, "onAdOpened")
            }

            override fun onAdClicked() {
                Log.d(TAG, "onAdClicked")
            }

            override fun onAdClosed() {
                Log.d(TAG, "onAdClosed")
            }
        }

        adManagerAdView!!.appEventListener = this
        adView.addView(adManagerAdView)

        gnWrapperAdBanner = GNWrapperAdBanner(this)
        gnWrapperAdBanner!!.setGnWrapperAdBannerListener(this)
        adView.addView(gnWrapperAdBanner)


        firebaseRemoteConfig = FirebaseRemoteConfig.getInstance()
        val configSettings = FirebaseRemoteConfigSettings.Builder()
                .setMinimumFetchIntervalInSeconds(3600)
                .build()
        firebaseRemoteConfig!!.setConfigSettingsAsync(configSettings)
        firebaseRemoteConfig!!.setDefaultsAsync(R.xml.remote_config_defaults)
        firebaseRemoteConfig!!.fetchAndActivate()
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        Log.d(TAG, "FirebaseRemoteConfig is Success")
                        var gnGNRemoteConfigValue = firebaseRemoteConfig!!.getString(GN_WRAPPER_CONFIG)
                        Log.d(TAG, "gnGNRemoteConfigValue: $gnGNRemoteConfigValue")

                        gnGNRemoteConfigValue = "{\"unit_id\":\"/9116787/1385323\",\"timeout\":0,\"is_refresh\":false,\"refresh_interval\":0,\"use_upr\":false,\"upr_settings\":{},\"use_hb\":false,\"hb_list\":[]}"


                        // Ad Load
                        gnWrapperAdBanner!!.load(gnGNRemoteConfigValue)
                    } else {
                        Log.w(TAG, "FirebaseRemoteConfig is failed Exception: " + task.exception)
                    }
                }
    }

    override fun onComplete(adUnitId: String?, gnCustomTargetingParams: GNCustomTargetingParams) {
        if (adManagerAdView!!.adUnitId == null) {
            adManagerAdView!!.adUnitId = adUnitId
            adManagerAdView!!.setAdSizes(AdSize.BANNER)
        }
        val adRequestBuilder = AdManagerAdRequest.Builder()
        for (key in gnCustomTargetingParams.keys) {
            Log.d(TAG, "key: " + key + " value: " + gnCustomTargetingParams[key])
            adRequestBuilder.addCustomTargeting(key, gnCustomTargetingParams[key])
        }
        val adRequest = adRequestBuilder.build()
        adManagerAdView!!.loadAd(adRequest)
    }

    override fun onError(gnErrorCode: GNErrorCode) {
        Log.w(TAG, "GNWrapperError: " + gnErrorCode.message)
    }

    override fun onAppEvent(s: String, s1: String) {
        if (s == "pubmaticdm") {
            Log.d(TAG, " GNWrapperAd.isShowInSDKView: " + gnWrapperAdBanner!!.isShowInSDKView)
            if (gnWrapperAdBanner!!.isShowInSDKView) {
                adManagerAdView!!.visibility = View.INVISIBLE
                gnWrapperAdBanner!!.show()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (adManagerAdView != null) {
            adManagerAdView!!.destroy()
        }
        if (gnWrapperAdBanner != null) {
            gnWrapperAdBanner!!.destroy()
        }
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val GN_WRAPPER_CONFIG = "GNWrapperConfig_Android"
    }
}
